librz/bin/p/bin_java.c

```diff
@@ -1,227 +1,259 @@
+	RzBinJavaClass *jclass = rz_bin_file_get_java_class(bf);
+	if (!jclass) {
+		return NULL;
+	}
+	RzBinInfo *binfo = RZ_NEW0(RzBinInfo);
+	if (!binfo) {
+		return NULL;
+	}
+	binfo->lang /*      */ = rz_bin_java_class_language(jclass);
+	binfo->file /*      */ = strdup(bf->file);
+	binfo->type /*      */ = strdup("JAVA CLASS");
+	binfo->bclass /*    */ = rz_bin_java_class_version(jclass);
+	binfo->has_va /*    */ = false;
+	binfo->rclass /*    */ = strdup("class");
+	binfo->os /*        */ = strdup("any");
+	binfo->subsystem /* */ = strdup("any");
+	binfo->machine /*   */ = strdup("jvm");
+	binfo->arch /*      */ = strdup("java");
+	binfo->bits /*      */ = 32;
+	binfo->big_endian /**/ = true;
+	binfo->dbg_info /*  */ = rz_bin_java_class_debug_info(jclass);
+	return binfo;
+}
 
-static int init(void *user) {
-	IFDBG_BIN_JAVA eprintf("Calling plugin init = %d.\n", DB ? 1 : 0);
-	if (!DB) {
-		IFDBG_BIN_JAVA eprintf("plugin DB beeing initted.\n");
-		DB = sdb_new("bin.java", NULL, 0);
-	} else {
-		IFDBG_BIN_JAVA eprintf("plugin DB already initted.\n");
+static bool load_buffer(RzBinFile *bf, void **bin_obj, RzBuffer *buf, ut64 loadaddr, Sdb *sdb) {
+	RzBinJavaClass *jclass = rz_bin_java_class_new(buf, loadaddr, sdb);
+	if (!jclass) {
+		return false;
 	}
-	return 0;
+	*bin_obj = jclass;
+	return true;
+}
+
+static void destroy(RzBinFile *bf) {
+	rz_bin_java_class_free(rz_bin_file_get_java_class(bf));
 }
 
-static int fini(void *user) {
-	IFDBG_BIN_JAVA eprintf("Calling plugin fini = %d.\n", DB ? 1 : 0);
-	if (!DB) {
-		IFDBG_BIN_JAVA eprintf("plugin DB already uninited.\n");
-	} else {
-		IFDBG_BIN_JAVA eprintf("plugin DB beeing uninited.\n");
-		sdb_free(DB);
-		DB = NULL;
+static bool check_buffer(RzBuffer *b) {
+	if (rz_buf_size(b) > 32) {
+		ut8 buf[4];
+		rz_buf_read_at(b, 0, buf, sizeof(buf));
+		return !memcmp(buf, "\xca\xfe\xba\xbe", 4);
 	}
+	return false;
+}
+
+static ut64 baddr(RzBinFile *bf) {
 	return 0;
 }
 
-static int add_sdb_bin_obj(const char *key, RzBinJavaObj *bin_obj) {
-	int result = false;
-	char *addr, value[1024] = { 0 };
-	addr = sdb_itoa((ut64)(size_t)bin_obj, value, 16);
-	if (key && bin_obj && DB) {
-		IFDBG_BIN_JAVA eprintf("Adding %s:%s to the bin_objs db\n", key, addr);
-		sdb_set(DB, key, addr, 0);
-		result = true;
-	}
-	return result;
+static Sdb *get_sdb(RzBinFile *bf) {
+	return bf->sdb;
 }
 
-static void add_bin_obj_to_sdb(RzBinJavaObj *bin) {
-	if (!bin) {
-		return;
+static void free_rz_bin_class(void /*RzBinClass*/ *k) {
+	RzBinClass *bclass = (RzBinClass *)k;
+	if (bclass) {
+		rz_list_free(bclass->methods);
+		rz_list_free(bclass->fields);
+		free(bclass->name);
+		free(bclass->super);
+		free(bclass->visibility_str);
+		free(bclass);
 	}
-	char *jvcname = rz_bin_java_build_obj_key(bin);
-	add_sdb_bin_obj(jvcname, bin);
-	bin->AllJavaBinObjs = DB;
-	free(jvcname);
 }
 
-static Sdb *get_sdb(RzBinFile *bf) {
-	RzBinObject *o = bf->o;
-	RzBinJavaObj *bin;
-	if (!o) {
+static RzList *classes(RzBinFile *bf) {
```

Comment:
there is no way to get two classes in a single file, is there?

Reply:
no